### DISCUSSION TITLE
How to get started?

### DISCUSSION THREAD

**Discussion by stevempotter** [terjeio/ioSender#116] [Q&A] [answered]
I am using the Teensy4.1 with USB and Ethernet connections. I have the grblHAL breakout board T41U5XBB v2.09. I have jumpered the Halt and Door lines to ground. But when I try to connect to IO Sender I get an Estop Active alert that will not let me proceed to get this set up. Is there another line I need to jumper to ground? 
Once it is connected, how will I know? 
How do I connect using the ethernet port?
Sorry for such basic questions. Perhaps more advice for newbies and a link to the built executable should be included in the Setup and Configuration page.

**Comment by terjeio** [comment]
>  But when I try to connect to IO Sender I get an Estop Active alert that will not let me proceed to get this set up. Is there another line I need to jumper to ground?

Perhaps the Halt input is configured for inversion? Try with the ground line removed. If this does not help connect with a terminal (such as putty) and check the real-time report response (press `?` to get it). If the `Pn:` element contains an `E` regardless of the Halt input beeing grounded or not check that you have the correct board map enabled in _my_machine.h_.

> Once it is connected, how will I know?

The main window will open and the DRO will contain values - likely 0. If blank then there is no controller connected or it has crashed. Since you get the estop message it is there and is running,

> How do I connect using the ethernet port?

Select the _Network_ tab in the connect dialog and fill in the IP address of the controller.

![image](https://user-images.githubusercontent.com/20260062/124775848-3f211b00-df3f-11eb-8c64-b80b2b4c7f50.png)

You can find the IP address (if assigned via DHCP) in the sender via _Help > About_ or by issuing a `$I` command in a terminal window.

![image](https://user-images.githubusercontent.com/20260062/124776384-b191fb00-df3f-11eb-9acf-4eef6c5994c5.png)

>  Perhaps more advice for newbies and a link to the built executable should be included in the Setup and Configuration page.

I'll add a link there to the [releases page](https://github.com/terjeio/ioSender/releases).

**Reply by stevempotter** [reply]
Thanks for your quick help, Terje! Yes, that was it: by removing the jumper from Halt to ground, I was able to start the IOsender with a USB interface, and see numbers in the DRO. USB Connection success!!!

And thanks to Phil Also, I dug through my collection of old routers and found one that was not locked down with proprietary firmware and followed the instructions in your manual. 
TELNET STREAM ACTIVE!!!!!
I set the IP address to manual, not DHCP so it won't change.
Do I need to keep using the router, now that I have established a connection? It would be much nicer to just plug an ethernet cable between the IOsender computer and the grblHAL board.
(My router is huge and clunky with 4 antennas!)

**Reply by phil-barrett** [reply]
Your router should allow reserving a specific IP address.

You can connect directly between a computer and the breakout board if you have an unused ethernet port on the PC.  However, you will need to assign a specific IP address that is not in the range managed by your router.  

You will need to set up the ethernet connection on the PC (start->settings->ethernet->change adapter options, right click Ethernet Network you want to change and select Properties->internet protocol Version 4->Properties) and set the the IP address on the breakout board (via Grbl Settings in ioSender).  Both will need to be in the same class C subnet.  The way I do it is I pick an address space that is close to my local/private space. If my local router space is 192.168.5.xxx, I will pick 192.168.6.xxx for the breakout board subnet.  PC would be 192.168.6.1 and BOB would be 192.168.6.2.  Gateway is 192.168.6.1 and subnet mask is 255.255.255.0  This is mostly from memory for Windows 10 so if there is any confusion, please don't hesitate to ask questions.

Note, you probably will not need a crossover ethernet cable unless you have a fairly old PC.

**Reply by stevempotter** [reply]
Thanks again, Phil. That worked on the first try! My CNC PC is a Lenovo Thinkpad T430, not too ancient, so no crossover cable needed.

**Comment by phil-barrett** [comment, marked as answer]
Setting up Ethernet is described in the [T41U5XBB Manual.](https://github.com/phil-barrett/grblHAL-teensy-4.x/blob/master/T41U5XBB%20v207.pdf)  It uses an older version of ioSender (nee GCodeSender) but it works the same.

**Reply by stevempotter** [reply]
I am having a great time using my new Teensy BOB and IOsender today, seeing how fast I can move my 1.5kW spindle up and down. I have managed to get it up to nearly 50mm/sec with no lost steps. This will really help with some of the z-intensive carving I do, such as terrain.

**Another basic IOsender question**: How does one enter a $ command? I typed it where you enter gcode commands, and nothing happened. Is there another place to send configuration commands within this program? I realize that in Grbl Settings tab, there are $ commands under the hood. But how to set them explicitly?

**Reply by raenji-sk** [reply]
There's a grbl:settings tab (or something like that) in the top left part of the screen.
You can configure all controller parameters there in an easy and intuitive way.
Much nicer than sending individual $ lines for each parameter.

**Reply by phil-barrett** [reply]
Yes, that's right.  

You can still enter $ commands in the MDI field in the lower left.  Perhaps you didn't click on the console tab?  That's where the results will be displayed.  And by the way, with newer grblHAL, $$  and $num doesn't output anything. Try $i.

**Reply by phil-barrett** [reply]
A bit more.  It is a bit complicated.  If you use another sender or a terminal app, $$, $n and $n=x work.

**Reply by terjeio** [reply]
`$$` output is normally claimed by the settings subsystem, go to _Help > About_ and click _To clipboard_ and paste into an editor for options, `$I`, `$N` and and `$$` output (including a brief description of each setting).

It _is_ possible to get the raw `$$` output to show up in the console as well, check the _Verbose_ and _Filter out realtime responses_ checkboxes before sending the `$$`-command from the MDI. 

`$n` and `$$=n` output (and output from  `$help` with or without arguments, e.g. `$help settings` lists all setting options with descriptions) is not formatted for sender use, it is formatted for human readability and intended for use when connected via a terminal such as putty. These commands should _not_ be issued from a sender MDI as the output may confuse the sender or even crash it (this since they are grblHAL extensions to the original Grbl protocol).

**Reply by stevempotter** [reply]
Thanks, I was confused both by sending commands that return nothing (? by itself and $$) and by not looking in the Console tab of IOsender. Including some basic instructions in Help: Usage Tips would be great for new users. What's there now assumes you already have the thing up and running and know where stuff is, or are already familiar with CNC senders. Same goes for the Wiki!
